### PR TITLE
Refactor `sanitize_mentions` method

### PR DESCRIPTION
**Pull request by rebelagentm**
Similar to https://github.com/dependabot/dependabot-core/pull/1558, we are refactoring this method to use the [commonmarker ](https://github.com/gjtorikian/commonmarker) gem.

### CODE REVIEW

File: common/spec/dependabot/pull_request_creator/message_builder_spec.rb [hunk 1]
```diff
@@ -1098,10 +1098,11 @@ def commits_details(base:, head:)
                 "<blockquote>\n"\
                 "<h2>v1.6.0</h2>\n"\
                 "<p>Mad props to <a href=\"https://github.com/greysteil\">"\
-                "@​greysteil</a> and <a href=\"https://github.com/hmarr\">"\
-                "@​hmarr</a> for the @angular/scope work - see "\
-                "<a href=\"https://github.com/gocardless/business/blob/HEAD/"\
-                "CHANGELOG.md\">changelog</a>.</p>\n"\
+                "@greysteil</a> and <a href=\"https://github.com/hmarr\">"\
```
Comment (rebelagentm) [inline common/spec/dependabot/pull_request_creator/message_builder_spec.rb]: @feelepxyz I feel like this HTML may be incorrect.  There are two hyperlinks for the hmarr mention.

**Comment by rebelagentm** [comment]
There's some Rubocop violations that I'm not sure if we should address or ignore.

File: common/lib/dependabot/pull_request_creator/message_builder/link_and_mention_sanitizer.rb [hunk 2]
```diff
@@ -29,6 +29,13 @@ class LinkAndMentionSanitizer
         CODEBLOCK_REGEX = /```|~~~/.freeze
```
Comment (feelepxyz) [inline common/lib/dependabot/pull_request_creator/message_builder/link_and_mention_sanitizer.rb]: Looks like this and the above comment block can be ✂️ 
```suggestion
```

File: common/lib/dependabot/pull_request_creator/message_builder/link_and_mention_sanitizer.rb [hunk 3]
```diff
@@ -37,49 +44,56 @@ def initialize(github_redirection_service:)
...
+          doc.to_html(COMMONMARKER_OPTIONS, COMMONMARKER_EXTENSIONS)
         end
 
         private
 
-        def sanitize_mentions(text)
-          text.gsub(%r{(?<![A-Za-z0-9`~])@#{GITHUB_USERNAME}/?}) do |mention|
-            next mention if mention.end_with?("/")
+        def sanitize_mentions(doc)
+          mention_regex = %r{(?<![A-Za-z0-9`~])@#{GITHUB_USERNAME}/?}
```
Comment (feelepxyz) [inline common/lib/dependabot/pull_request_creator/message_builder/link_and_mention_sanitizer.rb]: Could move this to the top of the class as `MENTION_REGEX`

File: common/lib/dependabot/pull_request_creator/message_builder/link_and_mention_sanitizer.rb [hunk 4]
```diff
@@ -37,49 +44,56 @@ def initialize(github_redirection_service:)
...
-            sanitized_mention = mention.gsub("@", "@&#8203;")
+          doc.walk do |node|
+            if node.type == :text && node.string_content.match?(mention_regex)
+              nodes = []
+              scan = StringScanner.new(node.string_content)
+
+              until scan.eos?
+                line = scan.scan_until(mention_regex) ||
+                       scan.scan_until(EOS_REGEX)
+                mention = line.match(mention_regex)&.to_s
```
Comment (feelepxyz) [inline common/lib/dependabot/pull_request_creator/message_builder/link_and_mention_sanitizer.rb]: Looks like there was a method on scan to get this
```suggestion
                mention = scan.matched
```

File: common/spec/dependabot/pull_request_creator/message_builder/link_and_mention_sanitizer_spec.rb [hunk 5]
```diff
@@ -57,7 +57,7 @@
 
         it "sanitizes the text with zero width space" do
           expect(sanitize_links_and_mentions).to eq(
-            "<p>[@​hmarr]</p>\n"
+            "<p>[<a href=\"https://github.com/hmarr\">@hmarr</a>]</p>\n"
```
Comment (feelepxyz) [inline common/spec/dependabot/pull_request_creator/message_builder/link_and_mention_sanitizer_spec.rb]: 👌

File: common/spec/dependabot/pull_request_creator/message_builder/link_and_mention_sanitizer_spec.rb [hunk 6]
```diff
@@ -171,8 +171,7 @@
           it "sanitizes the mention" do
             expect(sanitize_links_and_mentions).to eq(
               "<p><code>@command </code></p>\n<pre><code> @test~~~ "\
-              "[@&amp;#8203;feelepxyz](https://github.com/feelepxyz)\n</code>"\
-              "</pre>\n"
+              "@feelepxyz\n</code></pre>\n"
```
Comment (feelepxyz) [inline common/spec/dependabot/pull_request_creator/message_builder/link_and_mention_sanitizer_spec.rb]: 👌

File: common/spec/dependabot/pull_request_creator/message_builder_spec.rb (hunk 1 above)
Comment (feelepxyz) [reply common/spec/dependabot/pull_request_creator/message_builder_spec.rb]: Yes we've put a link in a link which I also think is invalid html 😬

Comment (feelepxyz) [reply common/spec/dependabot/pull_request_creator/message_builder_spec.rb]: Oh think I know what's gone wrong, I think we need to return early in the `walk` if the node is a `:link` so we don't end up walking inside the link and finding a text node there with a mention 😅

**Comment by feelepxyz** [comment]
@rebelagentm yeah can ignore those cops, going to relax the rules a bit because we keep hitting those and doesn't seem worth moving the code around to fit the rule.

There are also a couple of `pending` specs that should be fixed now with the refactor, just need to change them to `it`s.

**Comment by feelepxyz** [comment]
@rebelagentm oops fat-fingered the close button! Can you re-open the PR on your end?

**Comment by feelepxyz** [comment]
@rebelagentm merged it from a new pr instead so we can spot any issues before the holidays 🤠

**Pull request closed without merging**